### ISSUE TITLE
auth uids with periods break validation rules

### ISSUE THREAD

**Issue by dannycochran** [goldibex/targaryen#131]
".validate" rules are not respected when the auth uid has a period in it. For instance:

```javascript
const targaryen = require('targaryen');
// Note that "firstSignIn" must be a number
const rules = {
  rules: {
    ".write": true,
    ".read": true,
    users: {
      '$uid': {
        activity: {
          firstSignIn: {
            '.validate': 'newData.isNumber()'
          }
        }
      }
    }
  }
};
const user = { uid: 'foo.bar' };
const data = { users: { } };
const path = `/users/${user.uid}/activity/firstSignIn`;
const database = targaryen.database(rules, data).as(user).with({debug: true});
const result = database.write(path, 'notAnumber,shouldFail');
console.log(result.validated) // -> true
```

However if you change the above user to be "foobar", result.validated will correctly return "false".

I encountered this as I was creating dummy users using Math.random() for the uid, which is easy enough to work around, but it was a headache to track down why this is happening.

**Comment by dinoboff** [comment]
The problem is with dot in url:
```js
    it('should handle dot in url', function() {
      const rules = {
        rules: {
          '.write': true,
          '.read': true,
          $a: {
            '.validate': false
          }
        }
      };
      const db = database.create(rules, {}).with({debug: true});

      expect(db.write('foo', true).allowed).to.be.false();
      expect(db.write('foo.bar', true).allowed).to.be.false(); // fails
    });
```

**Comment by dinoboff** [comment]
Are dot allowed?

**Comment by dannycochran** [comment]
I don't think that's the problem. That's an ES6 template string, these two things are equivalent:

```javascript
const user = { uid: 'foo.bar' };
`/users/${user.uid}/activity/firstSignIn` === '/users/' + user.uid + '/activity/firstSignIn';
```

**Comment by dinoboff** [comment]
@dannycochran Firebase doesn't allow path containing dot.

**Comment by dannycochran** [comment]
Ah right, that makes sense. In which case, should targaryen throw an error that the path URL is invalid, rather than returning a result which indicates the write was valid?

**Comment by dinoboff** [comment]
It should throw. It must throw using an illegal location and it might need to validate the uid (the illegal character rule might apply to uid, I need to check).

**Comment by dinoboff** [comment]
The bug related to invalid path and node key is fixed in 3.0.3 and it's released.

Regarding UIDs, `firebase-admin` allow the creation of token with UID containing invalid characters, so targaryen is not validating them.

**Comment by chetbox** [comment]
This change breaks our tests and I believe it restricts a use case that is allowed in Firebase.

For safety, we user dummy UIDs that looks like email addresses to run our cloud functions with restricted privileges. Our write rules test paths of the form `/permissions/user@domain.com` which Firebase does allow us to do, but this version of targaryen does not allow such write rules.

Here's the error from our tests.

```
     AssertionError: Expected the write operation to succeed.
      
      Attempt to write /permissions/chet as {"uid":"user-admin@cloud-function.beeline.co"}.
      New Value: "null".
      
      /permissions/chet: write "newData.parent().parent().child('permissions').child(auth.uid).child('change_permissions').val() == true || auth.uid == 'user-admin' + '@cloud-function.beeline.co' && newData.val() == null"  => Error: Invalid location "permissions/user-admin@cloud-function.beeline.co/change_permissions" contains one of [., #, $, [, ]]
```

**Comment by dinoboff** [comment]
I will check with a proper write operation, but the simulator says "/permissions/user@domain.com" is an invalid location.

**Comment by dinoboff** [comment]
@chetbox with:

```js
'use strict';

const admin = require('firebase-admin');

const uid = 'foo@example.com';
const adminApp = admin.initializeApp({
  databaseURL: 'https://targaryen-42380.firebaseio.com',
  credential: admin.credential.cert('./sa.json')
});
const fooApp = admin.initializeApp({
  databaseURL: 'https://targaryen-42380.firebaseio.com',
  credential: admin.credential.cert('./sa.json'),
  databaseAuthVariableOverride: {uid}
}, 'foo');

init().then(write).then(close).catch(close);

function init() {
  const db = adminApp.database()

  return db.ref('/').set({admin: {[uid]: true}});
}

function write() {
  const db = fooApp.database()

  return db.ref('/foo').set('bar');
}

function clode(err) {
  if (err !== null) {
    console.error(err);
  }

  adminApp.delete();
  fooApp.delete();
}
```

I get:
```
node_modules/firebase-admin/lib/database/database.js:92
!0,!hd(b)))throw Error(a+" contains an invalid key ("+b+") "+Ec(c)+'.  Keys must be non-empty strings and can\'t contain ".", "#", "$", "/","[", or "]"');c.push(b);kd(a,g,c);c.pop()});if(d&&e)throw Error(a+' contains ".value" child '+Ec(c)+" in addition to actual children.");}}
           ^

Error: Firebase.set failed: First argument  contains an invalid key (foo@example.com) in property 'admin'.  Keys must be non-empty strings and can't contain ".", "#", "$", "/", "[", or "]"
    at Error (native)
    at node_modules/firebase-admin/lib/database/database.js:92:18
    at gb (node_modules/firebase-admin/lib/database/database.js:57:630)
    at kd (node_modules/firebase-admin/lib/database/database.js:91:424)
    at node_modules/firebase-admin/lib/database/database.js:92:166
    at gb (node_modules/firebase-admin/lib/database/database.js:57:630)
    at kd (node_modules/firebase-admin/lib/database/database.js:91:424)
    at jd (node_modules/firebase-admin/lib/database/database.js:90:876)
    at W.h.set (node_modules/firebase-admin/lib/database/database.js:258:241)
    at init (index.js:23:22)
```

**Comment by dinoboff** [comment]
@chetbox it should be fixed in [v3.0.5](https://github.com/goldibex/targaryen/releases/tag/v3.0.5)
